packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -17,81 +19,231 @@ import {
+// This is a slight variation of above, but we know the data is not null.
+type DataProcessor<T> = (data: NonNullable<T>) => T
+
+/**
+ * Create an update handler for the cache matrix. This updater will implicitly clone and assert
+ * that the data being passed to is is not undefined.
+ *  
+ * @param update 
+ * @returns 
+ */
+const createUpdateFunction =
+    <T>(update: DataProcessor<T>): Updater<T> =>
+    (data) =>
+        !data ? data : update(clone(data))
 
 export const cacheUpdateMatrix: CacheUpdateMatrix<Client> = {
     createCustomerAddress(customerId, {parameters}, response) {
         // getCustomerAddress uses `addressName` rather than `addressId`
         const newParams = {...parameters, addressName: response.addressId}
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            ...invalidateCustomer(newParams),
-            update: [{queryKey: getCustomerAddress.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerAddress.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        // Push new address onto the end of addresses list.
+                        customer?.addresses?.push(response)
+
+                        return customer
+                    })
+                }
+            ]
         }
     },
     createCustomerPaymentInstrument(customerId, {parameters}, response) {
         const newParams = {...parameters, paymentInstrumentId: response.paymentInstrumentId}
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            ...invalidateCustomer(newParams),
-            update: [{queryKey: getCustomerPaymentInstrument.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerPaymentInstrument.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        // Push new address onto the end of addresses list.
+                        customer?.paymentInstruments?.push(response)
+
+                        return customer
+                    })
+                }
+            ]
         }
     },
     createCustomerProductList(customerId, {parameters}, response) {
         // We always invalidate, because even without an ID we assume that something has changed
-        // TODO: Rather than invalidate, can we selectively update?
-        const invalidate: CacheUpdate['invalidate'] = [
-            {queryKey: getCustomerProductLists.queryKey(parameters)}
-        ]
-        // We can only update cache for this product list if we have the ID
         const listId = response.id
-        if (!listId) return {invalidate}
```

Comment:
I wasn't sure about why this id value might not exist. @kevinxh can you shed some light on that.

Reply:
let's remove it, I can't think of a reason either.

Reply:
The RAML spec (and therefore TypeScript type) does not guarantee that an `id` will be given. You'll have to ask the API team why that's the case, though.